### DISCUSSION TITLE
Input / search event listeners?

### DISCUSSION THREAD

**Discussion by ZachSaucier** [Pagefind/pagefind#535] [General]
Is there a way to add event listeners for when a user updates the search? I'm trying to update the query param in the search URL every time someone is typing.

Right now we can approximate it by doing something like this:

```js
const pagefind = new PagefindUI({ element: '#search' });
const search_params = new URLSearchParams(window.location.search);
pagefind.triggerSearch(search_params.get('q'));
const search_input = document.querySelector('#search input');
search_input.addEventListener('input', (e) => {
  search_params.set('q', e.target.value);
  const new_url = `${window.location.pathname}?${search_params.toString()}`;
  history.pushState(null, '', new_url);
});
```

But this has drawbacks:

1) Cleaning up the listener if the component is removed is more difficult.
2) It doesn't use the built in throttling that pagefind has.

Is there a way to do this better using a pagefind method? If not, does it make sense to add?

**Comment by bglw** [comment]
👋 hey @ZachSaucier 

One of my focuses once the new year rolls around and I get back into gear will be reworking and releasing Pagefind's modular UI, which already exists in its proto-form here: https://www.npmjs.com/package/@pagefind/modular-ui

This UI system has far more hooks and customization, already offering listeners on the core UI instance. This is the direction I'm wanting to head, with a goal that the modular UI replaces the current UI (by doing everything it does, but better). It isn't at that level yet, but that's the goal — so that's where I'm keen to spend any new energy 🙂

**Reply by bglw** [reply]
_(in other words, yes I think it's useful, but I'm personally unlikely to spend the time implementing it in the current UI since that time is better spent polishing the upcoming UI)_

**Reply by ZachSaucier** [reply]
Great, that makes sense to me! Shall I leave this discussion open until then or close it now?

**Reply by bglw** [reply]
Useful to keep this open so it stays on the mind 🙂
